stablehlo/transforms/StablehloProbe.cpp

```diff
@@ -0,0 +1,100 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <string>
+
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Support/LLVM.h"
+#include "stablehlo/dialect/StablehloOps.h"
+#include "stablehlo/reference/InterpreterOps.h"
+#include "stablehlo/transforms/Passes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+#define GEN_PASS_DEF_STABLEHLOPROBEPASS
+#include "stablehlo/transforms/Passes.h.inc"
+
+namespace {
+
+class StablehloProbePass
+    : public impl::StablehloProbePassBase<StablehloProbePass> {
+ public:
+  explicit StablehloProbePass() = default;
+  void runOnOperation() override;
+
+ private:
+  // Create a uniquely identifying probe ID for the given value being
+  // instrumented. Attempt to use named MLIR location data, otherwise use an
+  // increasing ID.
+  std::string getLocationNameOrUniqueId(Location location, unsigned int& id);
+
+  // Instrument a specified operation by adding an `interpreter.probe` op for
+  // each result produced by the operation.
+  void probeOp(Operation& op, unsigned int& id, OpBuilder& builder);
+
+  // Determine if a given operation is suitable for instrumentation. A suitable
+  // operation is defined as any operation which is not a ConstantOp, as such
+  // ops are effectively a no-op in terms of computation.
+  bool shouldProbeOp(Operation& op) const;
+};
+
+std::string StablehloProbePass::getLocationNameOrUniqueId(Location location,
+                                                          unsigned int& id) {
+  if (auto namedLocation = location.dyn_cast<NameLoc>())
+    return namedLocation.getName().strref().split('@').first.str();
+
+  return "probe" + std::to_string(++id);
+}
+
+void StablehloProbePass::probeOp(Operation& op, unsigned int& id,
+                                 OpBuilder& builder) {
+  for (Value value : op.getResults()) {
+    builder.setInsertionPointAfterValue(value);
+    Value instrumentedValue = builder.create<interpreter::ProbeOp>(
+        value.getLoc(), value,
+        StringAttr::get(&getContext(),
+                        getLocationNameOrUniqueId(value.getLoc(), id)));
+    value.replaceAllUsesExcept(instrumentedValue,
+                               instrumentedValue.getDefiningOp());
+  }
+}
+
+void StablehloProbePass::runOnOperation() {
+  ModuleOp module = getOperation();
+  OpBuilder builder(module);
+
+  // Strictly increasing counter to uniquely identify probe operations when MLIR
+  // location data is not available.
+  unsigned int probeId = 0;
+
+  module.walk([&](Operation* op) {
+    if (shouldProbeOp(*op)) probeOp(*op, probeId, builder);
```

Comment:
I had something like this in mind, to avoid the need to share anything by reference, also makes it clear where the value of ID is incremented:

```
if (shouldProbeOp(*op)) {
  for (auto res : op.getResults()) probeValue(res, ++id, builder);
}
```

Reply:
Also I'm wondering if we need something else, like:

```
if (!shouldProbeOp(*op)) return WalkResult::advance();
for (auto res : op.getResults()) {
  if (shouldProbeValue(res)) probeValue(res, ++id, builder);
}
```

I believe we only want to probe a result if the type is tensor type. I believe pass will produce invalid IR if the type is `!stablehlo.token`, since probe requires tensor operands?

Reply:
> I had something like this in mind, to avoid the need to share anything by reference, also makes it clear where the value of ID is incremented:

I think this will unconditionally increment `id` even if location data is used, slightly tweaked to only increment if location data isn't used.

I also updated additional checks for `TensorType` and updated the LIT test.

Reply:
> I think this will unconditionally increment id even if location data is used, slightly tweaked to only increment if location data isn't used.

Is this a bad thing? Should ID be appended to location name as well? I.e. `my_op.1` - it seems plausible that location metadata could use the same `opName` in `opName@something` (given that debug info is whatever you want it to be), leading to non-unique IDs

Reply:
That's a good point. There could be cases where the debug location in the original model leads to split ops that share the same `NamedLoc`. Updated the pass description and pass to do the following:

If MLIR location data is to be used (i.e. the `useDebugInfo` pass option is set), the `probe_id` will be computed as follows:

```
probe_id = x.#
```
Where `x` is `NamedLoc(x@...)` and `#` is a positive increasing integer. Otherwise, the probe will be named as `probe#`. I decided to leave an interleaving `.` between the location name and ID to make it easy to extract the named location from the suffix later on if needed.